lightning/backend/local.go

```diff
@@ -0,0 +1,1159 @@
+}
+
+type local struct {
+	engines  sync.Map
+	grpcClis grpcClis
+	splitCli split.SplitClient
+	tls  *common.TLS
+	pdAddr 	string
+
+	localStoreDir   string
+	regionSplitSize int64
+
+	rangeConcurrency  *worker.Pool
+	batchWriteKVPairs int
+	checkpointEnabled bool
+}
+
+// NewLocal creates new connections to tikv.
+func NewLocalBackend(
+	ctx context.Context,
+	tls *common.TLS,
+	pdAddr string,
+	regionSplitSize int64,
+	localFile string,
+	rangeConcurrency int,
+	sendKVPairs int,
+	enableCheckpoint bool,
+) (Backend, error) {
+	pdCli, err := pd.NewClient([]string{pdAddr}, tls.ToPDSecurityOption())
+	if err != nil {
+		return MakeBackend(nil), errors.Annotate(err, "construct pd client failed")
+	}
+	splitCli := split.NewSplitClient(pdCli, tls.TransToTlsConfig())
+
+	shouldCreate := true
+	if enableCheckpoint {
+		if info, err := os.Stat(localFile); err != nil {
+			if !os.IsNotExist(err) {
+				return MakeBackend(nil), err
+			}
+		} else if info.IsDir() {
+			shouldCreate = false
+		}
+	}
+
+	if shouldCreate {
+		err = os.Mkdir(localFile, 0700)
+		if err != nil {
+			return MakeBackend(nil), err
+		}
+	}
+
+	local := &local{
+		engines:  sync.Map{},
+		splitCli: splitCli,
+		tls:  	  tls,
+		pdAddr:pdAddr,
+
+		localStoreDir:   localFile,
+		regionSplitSize: regionSplitSize,
+
+		rangeConcurrency:  worker.NewPool(ctx, rangeConcurrency, "range"),
+		batchWriteKVPairs: sendKVPairs,
+		checkpointEnabled: enableCheckpoint,
+	}
+	local.grpcClis.clis = make(map[uint64]*grpc.ClientConn)
+	return MakeBackend(local), nil
+}
+
+func (local *local) getGrpcConnLocked(ctx context.Context, storeID uint64) (*grpc.ClientConn, error) {
+
+	store, err := local.splitCli.GetStore(ctx, storeID)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	opt := grpc.WithInsecure()
+	if local.tls.TransToTlsConfig() != nil {
+		opt = grpc.WithTransportCredentials(credentials.NewTLS(local.tls.TransToTlsConfig()))
+	}
+	ctx, cancel := context.WithTimeout(ctx, dialTimeout)
+	keepAlive := 10
+	keepAliveTimeout := 3
+	bfConf := backoff.DefaultConfig
+	bfConf.MaxDelay = time.Second * 3
+	conn, err := grpc.DialContext(
+		ctx,
+		store.GetAddress(),
+		opt,
+		grpc.WithConnectParams(grpc.ConnectParams{Backoff: bfConf}),
+		grpc.WithKeepaliveParams(keepalive.ClientParameters{
+			Time:                time.Duration(keepAlive) * time.Second,
+			Timeout:             time.Duration(keepAliveTimeout) * time.Second,
+			PermitWithoutStream: true,
+		}),
+	)
+	cancel()
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	// Cache the conn.
+	local.grpcClis.clis[storeID] = conn
+	return conn, nil
+}
+
+// Close the importer connection.
+func (local *local) Close() {
+	local.engines.Range(func(k, v interface{}) bool {
+		v.(*LocalFile).Close()
+		return true
+	})
+
+	if !local.checkpointEnabled || common.IsEmptyDir(local.localStoreDir) {
+		err := os.RemoveAll(local.localStoreDir)
+		if err != nil {
+			log.L().Error("remove local db file failed", zap.Error(err))
+		}
+	}
```

Comment:
```suggestion
	if !local.checkpointEnabled {
		err := os.Remove(local.localStoreDir)
		if err != nil {
			log.L().Warn("remove local db file failed", zap.String("dir", local.localStoreDir), log.ShortError(err))
		}
	}
```

* Don't use Error level for non-fatal errors.
* Is it normal that we close the local engine with left over files? I guess a warning in case of non-empty directory is fine.

Reply:
When lightning exits before finished (e.g. by user manually terminates), backend will be closed, but currently data in each engine hasn't been imported. So we can't deleted all the data file if checkpoint is enable.

Reply:
@glorv the question was about the case when the checkpoint is *disabled*, where the other condition `common.IsEmptyDir(local.localStoreDir)` was triggered.

Reply:
The dir is empty if lightning is finished successfully, thus all engine data in this directory is clean up, then this dir will be empty. We created this dir in `NewLocalBackend`, so we should remove here.